design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,187 @@
+# KLIP 15 - KQSLDB Client and new API
+
+**Author**: Tim Fox (purplefox) | 
+**Release Target**: ? | 
+**Status**: _In Discussion_ | 
+**Discussion**:
+
+Please read the initial discussion on the ksqlDB developer group first:
+
+https://groups.google.com/forum/#!topic/ksql-dev/yxcRlsOsNmo
+
+And there is a partially working prototype for this work described here:
+
+https://groups.google.com/forum/#!topic/ksql-dev/5mLKvtZFs4Y
+
+ For ksqlDB to be a successful project we need to provide an awesome out of the box developer experience
+ and it should be super easy and fun to write powerful event streaming applications.
+ Currently this is somewhat diffificult to do, and it requires the use of multiple clients and multiple moving parts in order to build
+ a simple event sourcing / CQRS style application.
+ We should provide a new ksqlDB client that provides that currently missing single interaction point with the event streaming
+ platform so that we can improve the application development process.
+           
+## Motivation and background
+
+In order to increase adoption of ksqlDB we need the developer experience to be as slick as possible. Currently, in order to write
+a simple event streaming application with ksqlDB, up to 4 clients may be needed: 
+
+1. An HTTP client to access the current HTTP/REST API
+2. A WebSocket client to access the streaming query WebSockets endpoint
+3. A JDBC client to access data in JDBC tables that are synced with KSQL via connect, as the current support for
+pull queries does not make the data otherwise easily accessible.
+4. A Kafka client for producing and consuming messages from topics.
+
+This is a lot for an application to handle and creates a steep learning curve for application developers.
+
+Moreover, our current HTTP API is lacking, especially in the area of streaming query results which makes it difficult to write an application
+that uses this API.
+
+This KLIP proposes that we:
+
+* Create a ksqlDB client (initially in Java and JavaScript) as the primary interaction point to the event streaming platform
+for an application
+* Create a new server side implementation that supports the client and enables the functionality of the client to be delivered in a straightforward
+and efficient way
+* The creation of a new wire protocol for the client<->server communication.
+* Migrate the functionality of the current HTTP/REST API over to the new server implementation and retire the parts that are no longer needed.
+
+## What is in scope
+
+### The client
```

Comment:
@purplefox can you include some mock/example client code that illustrates how using a client might look and feel?

Reply:
Hey @derekjn - I put together a prototype which shows a real(ish) client, examples, and an example app. Links are on the post here https://groups.google.com/forum/#!topic/ksql-dev/5mLKvtZFs4Y